feat: Adds the /explore endpoint to the v1 API

michael-s-molina opened the pull request:
### SUMMARY
This PR is part of the Single Page App initiative and it adds the `/explore` endpoint to the `v1` API. The previous legacy endpoint in `superset.views.core` was responsible not only for assembling the Explore context but also to save Explore-related information. This endpoint, however, does not contain any save operation. We should use the existing `v1` endpoints to save the charts and dashboard relationships. Another important difference is that this endpoint does not return shared bootstrap data such as `user`, `permissions`, `common`, etc. as these are already available by other means.

The legacy endpoint also contained redirects and flash messages. These were transformed into HTTP response errors/values and should be handled by the client-side.

Much of the code in the new endpoint comes from the legacy one. It's important to mention that it is not the objective of the PR to refactor the code at this point to avoid introducing potential regressions. 

```
GET /v1/explore/

QUERY PARAMS # all optional
   form_data_key=<key>
   permalink_key=<key>
   slice_id=<id>
   dataset_id=<id>
   dataset_type=<type> 

RESPONSE 
{ 
   form_data: <object>,
   dataset: <object>,
   slice: <object>,
   message: <string> # information related to the request
}

```
We will remove the legacy endpoint after the SPA initiative is completed.

### TESTING INSTRUCTIONS
1 - Execute the API tests
2 - All tests should pass

### ADDITIONAL INFORMATION
- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [x] Introduces new feature or API
- [ ] Removes existing feature or API

michael-s-molina committed:
feat: Adds the /explore endpoint to the v1 API

michael-s-molina committed:
Fixes pylint errors

michael-s-molina committed:
Fixes tests

michael-s-molina committed:
Changes the tests logic

zhaoyongjie commented on superset/explore/api.py:
```diff
@@ -0,0 +1,127 @@
+logger = logging.getLogger(__name__)
+
+
+class ExploreRestApi(BaseApi, ABC):
```
Why abstract class here? I don't see any classes that inherit from this abstract class.

zhaoyongjie commented on superset/explore/commands/get.py:
```diff
@@ -0,0 +1,172 @@
+    def run(self) -> Optional[Dict[str, Any]]:
+        initial_form_data = {}
+
+        if self._permalink_key is not None:
+            command = GetExplorePermalinkCommand(self._actor, self._permalink_key)
+            permalink_value = command.run()
+            if permalink_value:
+                state = permalink_value["state"]
+                initial_form_data = state["formData"]
+                url_params = state.get("urlParams")
+                if url_params:
+                    initial_form_data["url_params"] = dict(url_params)
+            else:
+                raise ExplorePermalinkGetFailedError()
```
Please try to avoid nested `if else expression`

michael-s-molina replied:
Avoiding refactors at this stage 😉

ktmud replied:
I believe in this case you can easily raise the exception earlier and avoid an additional indentation.

michael-s-molina replied:
Ok, I added the suggestion. Just keep in mind that the code is pretty much a copy and paste of the old endpoint. We're just trying to rearrange things at this point so we can provide the same behavior through SPA.

zhaoyongjie commented on superset/explore/commands/get.py:
```diff
@@ -0,0 +1,172 @@
+                    initial_form_data["url_params"] = dict(url_params)
+            else:
+                raise ExplorePermalinkGetFailedError()
+        elif self._form_data_key:
+            parameters = FormDataCommandParameters(
+                actor=self._actor, key=self._form_data_key
+            )
+            value = GetFormDataCommand(parameters).run()
+            initial_form_data = json.loads(value) if value else {}
+
+        message = None
```
Many exceptions can be caught in the `self.validate` method.

michael-s-molina replied:
Avoiding refactors at this stage 😉

dpgaspar replied:
I agree with @zhaoyongjie

michael-s-molina replied:
I also agree with @zhaoyongjie. I was just trying to avoid refactors at this stage and preserve the old endpoint code but it seems it will not be possible :man_shrugging:. I'll implement the suggestion.

michael-s-molina replied:
Ok. I'll leave the refactor for a follow-up then. @zhaoyongjie commented that we may have parts that are not even being used anymore. After we complete the SPA work, we'll have more knowledge of the domain as well, which will give us more information for the refactoring.

zhaoyongjie reviewed (commented):
Thanks for improving explore endpoint, I have a first quick review and leave some comments.

michael-s-molina committed:
Removes ABC reference

michael-s-molina committed:
Improves indentation

ktmud commented on superset/explore/api.py:
```diff
@@ -0,0 +1,126 @@
+        except TemporaryCacheAccessDeniedError as ex:
+            return self.response(403, message=str(ex))
+        except TemporaryCacheResourceNotFoundError as ex:
+            return self.response(404, message=str(ex))
```
I really really believe these common errors should be handled in some kind of base class. We shouldn't have to duplicate the same code in every API endpoint.

michael-s-molina replied:
I agree. We even created a task in our backlog to do this work and adjust all the endpoints but it always loses the priority battle 😞. Do you think you can work on it?

kgabryje reviewed (approved):
LGTM!

dpgaspar reviewed (commented):
Looking good!

Can you add unittests for `GetExploreCommand` also

michael-s-molina committed:
Addresses review comments

michael-s-molina commented on superset/explore/api.py:
```diff
@@ -0,0 +1,126 @@
+                form_data_key=request.args.get("form_data_key"),
+                dataset_id=request.args.get("dataset_id"),
+                dataset_type=request.args.get("dataset_type"),
+                slice_id=request.args.get("slice_id"),
```
Nice catch! I added the type definition and catch the `ValidationError` with a `400` error.

michael-s-molina commented on superset/explore/schemas.py:
```diff
@@ -0,0 +1,80 @@
+    time_grain_sqla = fields.List(fields.List(fields.Str()))
+    type = fields.Str()
+    uid = fields.Str()
+    verbose_map = fields.Dict()
```
I added the descriptions as best as I could. Feel free to suggest any improvements to the texts.

michael-s-molina commented on superset/explore/schemas.py:
```diff
@@ -0,0 +1,80 @@
+    cache_timeout = fields.Int()
+    column_formats = fields.Dict()
+    columns = fields.List(fields.Dict())
+    database = fields.Dict()
```
Oh boy 😆. Some of them are dynamic and others have really extensive trees 😞. It's also hard to check all the possible values because many are defined as dictionaries. For this PR, I limited the depth to two levels and tried to map the attributes the best I could.

dpgaspar replied:
it's a pretty great improvement already

michael-s-molina commented on superset/explore/api.py:
```diff
@@ -0,0 +1,126 @@
+        """Retrives Explore initial context.
+        ---
+        get:
+          description: >-
```
Cool. I didn't know about the summary. Now it looks like this:

<img width="1138" alt="Screen Shot 2022-06-22 at 10 09 50 AM" src="https://user-images.githubusercontent.com/70410625/175045255-d94f911e-40c9-4498-a6c6-a58b65f54085.png">

dpgaspar replied:
nice!

michael-s-molina commented on superset/explore/api.py:
```diff
@@ -0,0 +1,126 @@
+    method_permission_name = MODEL_API_RW_METHOD_PERMISSION_MAP
+    include_route_methods = {RouteMethod.GET}
+    allow_browser_login = True
+    class_permission_name = "ExploreRestApi"
```
Done 😉

michael-s-molina commented:
Thanks for your review @dpgaspar. I really appreciate it!

I noticed that many concepts are shared among API schema definitions (database, form_data, dataset, etc). Currently, some are duplicated in each endpoint and are inconsistent with one another. One thing we could do in a follow-up is to create many of the schema concepts in a more global way, to share both structure and documentation. We should also provide the ability to export a concept partially like just exposing some specific properties of the global database definition. This would improve our API standardization a lot not only by ensuring an object is correctly defined but also that the documentation associated with that object is consistent among endpoints.

michael-s-molina committed:
Rebases code

michael-s-molina commented:
> Can you add unittests for GetExploreCommand also

Since the API is pretty much invoking the `GetExploreCommand`, with no extra logic, the tests for the command would be very similar. So I chose to just test the API layer because it seemed like the best cost/benefit approach.

dpgaspar commented:
> Thanks for your review @dpgaspar. I really appreciate it!
> 
> I noticed that many concepts are shared among API schema definitions (database, form_data, dataset, etc). Currently, some are duplicated in each endpoint and are inconsistent with one another. One thing we could do in a follow-up is to create many of the schema concepts in a more global way, to share both structure and documentation. We should also provide the ability to export a concept partially like just exposing some specific properties of the global database definition. This would improve our API standardization a lot not only by ensuring an object is correctly defined but also that the documentation associated with that object is consistent among endpoints.

sounds promising!

dpgaspar commented on tests/integration_tests/explore/api_tests.py:
```diff
@@ -0,0 +1,211 @@
+        == "Form data not found in cache, reverting to chart metadata."
+    )
+    assert result["slice"] != None
+    assert result["slice"] != None
```
duplicate assertion

dpgaspar commented:
> > Can you add unittests for GetExploreCommand also
> 
> Since the API is pretty much invoking the `GetExploreCommand`, with no extra logic, the tests for the command would be very similar. So I chose to just test the API layer because it seemed like the best cost/benefit approach.

I understand, my current concern is that the defined response payload is complex and I don't see them been asserted on the API tests, for example `dataset` and `slice` are always asserted has `None`. Can you add tests for them?

michael-s-molina commented:
> btw, we are currently adding this: https://github.com/apache/superset/blob/master/superset/views/core.py#L1352 to the migrated `/superset/*` endpoints, not actually deleting them, the reason being that actually deleting the endpoints is considered a breaking change. The plan is to delete all these on v3

Nice! I'll add this to the old explore endpoint as well.

michael-s-molina committed:
Improves dataset and slice assertions

michael-s-molina commented:
> I understand, my current concern is that the defined response payload is complex and I don't see them been asserted on the API tests, for example `dataset` and `slice` are always asserted has `None`. Can you add tests for them?

@dpgaspar I added the assertions to `dataset` and `slice`. Since they are huge trees, I partially matched using the most important attributes.

I also added the warning to the old explore endpoint.

michael-s-molina committed:
Fixes tests

michael-s-molina committed:
Removes schema and table name assertions

michael-s-molina committed:
Removes fixed IDs

michael-s-molina committed:
Fixes datasource ID
